### DISCUSSION TITLE
Token Information Dynamics Control (Early-Exit)

### DISCUSSION THREAD

**Discussion by dingste** [ggml-org/llama.cpp#18702] [Show and tell]
This feature implements dynamic inference control based on the **information dynamics** of the logits. It allows the model to prematurely terminate generation as soon as statistical convergence or excessive energy dissipation is detected.

### 1. Local Entropy Stability

Instead of simply waiting for the `<|end_of_text|>` token, this feature measures the **logit gap** (the difference between the most probable and the second most probable token). A high value indicates high prediction certainty (**sense prediction**).

### 2. Cross-Layer Refinement Rate & EMA

Changes in stability over time are smoothed using an **Exponential Moving Average (EMA)**. This filters out mathematical noise (especially in quantized models) and isolates the actual trend in information density.

### 3. Energy Accumulation

Every significant change in the prediction direction costs "energy." If the accumulated energy exceeds a threshold (**burnout**), the inference process terminates. This prevents the model from drifting into endless, redundant loops.

---

## Advantages Compared (Example Scenario)

*Based on a test setup: Mistral 7B, Q8_0, ~1000 Prompt Tokens.*

| Feature | Standard Inference | Early Exit |
| --- | --- | --- |
| **Termination** | EOS or Token Limit | C-Gap or S-Budget Limit |
| **Efficiency** | 100% Compute Load | **~30% - 50% Compute Load** |
| **Behavior** | Tends to redundancy ("yapping") | Concise, ends at the information peak |
| **Hallucination** | Risk increases with long output | **Minimized** by timely termination |

### The Counter-Caricature of Reasoning

While **reasoning methods** (such as Chain-of-Thought) increase computation time through *expansion* to enforce quality, **Early-Exit** is its reductive counterpart. It preserves quality through *selection* by stopping inference precisely before statistical noise becomes overwhelming.

---

### Parameters

* `--early-exit`: Enables the feature and disables potential reasoning.

* `--early-exit-gap N`: Threshold for stable jumps. (Default: `6.0`). Higher values ​​result in later termination with high certainty.
* `--early-exit-burnout N`: Limit for accumulated energy. (Default: 100.0). Protects against drifting and infinite loops.

### Example Command

```bash

./llama-server -m model.gguf --early-exit --early-exit-gap 12.0 --early-exit-burnout 200.0

```

### Visual Feedback

In the event of premature termination, the system injects a `...` into the stream. This signals to the user (or client) that the generation was successfully terminated by the Dynamics supervisor and that no technical fault has occurred.

---

## Technical Implementation

The check is performed immediately after logit generation and **before** the sampler chain. This ensures that the measurement is based on the model's unadulterated probabilities and saves computing power on unnecessary sampling steps.
